File: php/src/FactChecker.php
<?php

declare(strict_types=1);

namespace FactChecker;

use FactChecker\AssessService\Assessor;
use FactChecker\Exceptions\CannotFetchFact;
use FactChecker\FetchService\Fetcher;
use FactChecker\LoggerService\Logger;
use FactChecker\LoggerService\NullLogger;
use FactChecker\NotifyService\Notifier;
use FactChecker\NotifyService\NullNotifier;
use RuntimeException;

class FactChecker
{
    const API_URL = 'https://catfact.ninja/fact';

    private Logger $logger;

    private Notifier $notifier;

    private Fetcher $fetcher;

    private Assessor $assessor;

    public function __construct(Fetcher $fetcher, Assessor $assessor)
    {
        $this->initLogger();
        $this->initNotifier();

        $this->initFetcher($fetcher);
        $this->initAssessor($assessor);
    }

    private function initLogger(): void
    {
        $this->logger = new NullLogger();
    }

    private function initNotifier(): void
    {
        $this->notifier = new NullNotifier();
    }

    private function initFetcher(Fetcher $fetcher): void
    {
        $this->fetcher = $fetcher;
    }

    private function initAssessor(Assessor $assessor): void
    {
        $this->assessor = $assessor;
    }

    public function setLogger(Logger $logger): void
    {
        $this->logger = $logger;
    }

    public function setNotifier(Notifier $notifier): void
    {
        $this->notifier = $notifier;
    }

    /**
     * Return a fetched random fact with an assessment.
     *
     * @return string
     */
    public function randomFact(): string
    {
        try {
            $fact = $this->fetchFact();
        } catch (CannotFetchFact $e) {
            return sprintf('Cannot retrieve a fact due to an error: %s.', $e->getMessage());
        }

        $assessment = $this->assessFact($fact);

        return sprintf('%s %s', $fact, $assessment);
    }

    /**
     * Fetch a fact.
     *
     * @return string
     *
     * @throws CannotFetchFact
     */
    protected function fetchFact(): string
    {
        try {
            $rawFact = $this->fetcher->fetch(self::API_URL);
        } catch (RuntimeException $e) {
            // log special cases: - RequestError; - RequestTimeout; - etc.
            $this->logger->error($e->getMessage());
            throw new CannotFetchFact($e->getMessage());
        }

        try {
            $parsedFact = json_decode($rawFact, false, 2, JSON_THROW_ON_ERROR);

            if (!property_exists($parsedFact, "fact")) {
                throw new \LogicException(
                    sprintf('the fact field doesn\'t exist. The original JSON is: %s.', $rawFact)
                );
            }

            $fact = $parsedFact->fact;

        } catch (\JsonException $e) {
            // log special cases: - JSON parse exception
            $this->logger->alert($e->getMessage());
            throw new CannotFetchFact($e->getMessage());
        } catch (\LogicException $e) {
            // log special cases: - JSON is unexpected
            $this->logger->alert($e->getMessage());
            // notify special cases: - notify Discord, - notify Mail
            $this->notifier->notify('email', $e->getMessage(), ['to' => 'CTO']);
            $this->notifier->notify('email', $e->getMessage(), ['to' => 'programmers']);
            $this->notifier->notify('slack', $e->getMessage(), ['to' => 'programmers']);
            throw new CannotFetchFact($e->getMessage());
        }

        return $fact;
    }

    protected function assessFact(string $fact): string
    {
        $opinion = $this->assessor->getOpinion($fact);
        $score = $this->assessor->getScore($fact);

        return sprintf(
            'It seems to be %s. Our score is %s point%s.',
            $opinion,
            $score,
            ($score !== 1) ? 's' : '',
        );
    }
}

File: java/app/src/main/java/factchecker/FactChecker.java
package factchecker;

import factchecker.AssessService.Assessor;
import factchecker.FetchService.Fetcher;
import factchecker.LoggerService.Logger;
import factchecker.LoggerService.NullLogger;
import factchecker.NotifierService.Notifier;
import factchecker.NotifierService.NullNotifier;
import org.json.JSONException;
import org.json.JSONObject;

import javax.security.auth.login.LoginException;
import java.io.IOException;

public class FactChecker {
    public static final String API_URL = "https://catfact.ninja/fact";

    private Logger logger;

    private Notifier notifier;

    private Fetcher fetcher;

    private Assessor assessor;

    public FactChecker(Fetcher fetcher, Assessor assessor) {
        initLogger();
        initNotifier();

        initFetcher(fetcher);
        initAssessor(assessor);
    }

    private void initLogger() {
        this.logger = new NullLogger();
    }

    private void initNotifier() {
        this.notifier = new NullNotifier();
    }

    private void initFetcher(Fetcher fetcher) {
        this.fetcher = fetcher;
    }

    private void initAssessor(Assessor assessor) {
        this.assessor = assessor;
    }

    public void setLogger(Logger logger) {
        this.logger = logger;
    }

    public void setNotifier(Notifier notifier) {
        this.notifier = notifier;
    }

    public String randomFact() {
        String fact = "";

        try {
            fact = fetchFact();
        } catch (CannotFetchFact e) {
            return String.format("Cannot retrieve a fact due to an error: %s.", e.getMessage());
        }

        String assessment = assessFact(fact);

        return String.format("%s %s", fact, assessment);
    }

    protected String fetchFact() throws CannotFetchFact {
        String rawFact = "";
        String fact = "";

        try {
            rawFact = this.fetcher.fetch(API_URL);
        } catch (IOException | InterruptedException e) {
            // log special cases: - NoHttpResponse; - ConnectTimeoutException; - etc
            this.logger.error(e.getMessage());
            throw new CannotFetchFact(e.getMessage(), e);
        }

        try {
            JSONObject parsedFact = new JSONObject(rawFact);

            if (!parsedFact.has("fact")) {
                throw new LoginException(String.format("the fact field doesn't exist. The original JSON is: %s.", rawFact));
            }

            fact = parsedFact.getString("fact");
        } catch (JSONException e) {
            // log special cases: - JSON parse exception
            this.logger.fatal(e.getMessage());
            throw new CannotFetchFact(e.getMessage(), e);
        } catch (LoginException e) {
            // log special cases: - JSON is unexpected
            this.logger.fatal(e.getMessage());
            // notify special cases: - notify Discord, - notify Mail
            this.notifier.notify("email", e.getMessage());
            this.notifier.notify("email", e.getMessage());
            this.notifier.notify("slack", e.getMessage());
            throw new CannotFetchFact(e.getMessage(), e);
        }

        return fact;
    }

    protected String assessFact(String fact) {
        return generateAssessment(fact);
    }

    protected String generateAssessment(String fact) {
        String opinion = this.assessor.getOpinion(fact);
        int score = this.assessor.getScore(fact);

        return String.format(
            "It seems to be %s. Our score is %s point%s.",
            opinion,
            score,
            (score != 1) ? "s" : ""
        );
    }
}
